core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1357,19 +1382,61 @@ public void startRecovery(DiscoveryNode localNode, DiscoveryNode sourceNode, Rec
             }
         } else if (restoreSource == null) {
             // recover from filesystem store
+
+            IndexMetaData indexMetaData = indexSettings().getIndexMetaData();
+            Index mergeSourceIndex = indexMetaData.getMergeSourceIndex();
+            final boolean recoverFromLocalShards = mergeSourceIndex != null && shardRouting.allocatedPostIndexCreate(indexMetaData) == false && shardRouting.primary();
             final RecoveryState recoveryState = new RecoveryState(shardId(), shardRouting.primary(),
-                RecoveryState.Type.STORE, localNode, localNode);
-            markAsRecovering("from store", recoveryState); // mark the shard as recovering on the cluster state thread
-            threadPool.generic().execute(() -> {
-                try {
-                    if (recoverFromStore(localNode)) {
-                        recoveryListener.onRecoveryDone(recoveryState);
+                recoverFromLocalShards ? RecoveryState.Type.LOCAL_SHARDS : RecoveryState.Type.STORE, localNode, localNode);
+            if (recoverFromLocalShards) {
+                final List<IndexShard> startedShards = new ArrayList<>();
+                // look up if the index actually exists - we throw the right exception below instead of failing with ISE
+                final IndexMetaData sourceIndexMetaData = indicesService.getIndexMetaData(mergeSourceIndex);
+                final IndexService sourceIndexService = indicesService.indexService(mergeSourceIndex);
```

Comment:
1. we don't want to throw that exception - we want to pass it to the listener
2. we want only throw the exception if the index is not existing (not if it's just not allocated on this node)
3. there is no mismatch since we use UUID